### PR TITLE
[Elsa] Main assignment - Checking if two words are anagrams

### PR DESCRIPTION
**Pull request by elsatellite**
#### `WordAnagramChecker.check(A, B)` returns `false` if:
- A or B are `null`
- A or B contain non-letter characters
- There's no permutation of the characters of A that can produce B (case insensitive).

Is there a little bit of overkill in the tests?

### CODE REVIEW

File: elsatellite/assignment-1/src/test/WordAnagramCheckerTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    assertEquals(checker.check(word, sameWordButDifferentCase), true, "Same words but different case should be anagrams.");
+  }
+
+  @Test
+  public void permutationOfCharactersShouldBeValid() {
+    String word = "permuteme";
+    String permutedWord;
+
+    for (int i = 0 ; i < 1000 ; i++) {
```
Comment (elsatellite) [inline elsatellite/assignment-1/src/test/WordAnagramCheckerTest.java]: I agree this looks wacky. I think tests shouldn't have elements of randomness. However I think it's not that bad because, on one side, having one example like "ABC" vs "BAC" is a bit too specific and on the other side generating all the possible permutations is a bit too much. At least here I'm controlling the ressources.

File: elsatellite/assignment-1/src/main/AnagramChecker.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    return false;
+  }
+
+  /**
+   * Counts the occurence of each letter in a string.
+   *
+   * @param A a string
+   * @return  a map where a value is the occurence of its key in A.
+   */
+  private HashMap<Character, Integer> getLetterCount(String A) {
```
Comment (haixiaosu) [inline elsatellite/assignment-1/src/main/AnagramChecker.java]: para names should be lowercased in general.

File: elsatellite/assignment-1/src/main/AnagramChecker.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+   * @param A string to check
+   * @return  true if A is a word.
+   */
+  private boolean isNotAWord(String A) {
+    if (A == null || A.isEmpty()) {
+      return true;
+    }
+
+    for (int i = 0 ; i < A.length() ; i++) {
+      if (!Character.isLetter(A.charAt(i))) {
```
Comment (haixiaosu) [inline elsatellite/assignment-1/src/main/AnagramChecker.java]: Is there a reason you are checking whether the character is a letter?

File: elsatellite/assignment-1/src/main/AnagramChecker.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+   * @param A a string
+   * @param B a string
+   * @return  true if A and B are two words that are anagrams of each other.
+   */
+  public boolean check(String A, String B) {
+    if (isNotAWord(A) || isNotAWord(B)) {
+      return false;
+    }
+
+    A = A.toLowerCase();
```
Comment (haixiaosu) [inline elsatellite/assignment-1/src/main/AnagramChecker.java]: Why is this needed?

Comment (haixiaosu) [reply elsatellite/assignment-1/src/main/AnagramChecker.java]: Ah, I think the main assignment didn't specify that upper case should also equal to lower case. To my understanding, the assignment requires the solution to deal with only lowercase, it doesn't mean that 'A' == 'a' :)

File: elsatellite/assignment-1/src/test/WordAnagramCheckerTest.java (hunk 1 above)
Comment (haixiaosu) [reply elsatellite/assignment-1/src/test/WordAnagramCheckerTest.java]: You're correct that tests shouldn't have elements of randomness, and it's a good thing that this will cover all cases. However, the more complicated the test helper, the more bugs it might cause in the actual test. 

It's okay in the unit test to only test for some of the permutations. For example: ABC compared CBA compared ACB, there's no need to test for all permutations.

File: elsatellite/assignment-1/src/test/AnagramCheckerTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  @Test
+  public void sentencesOfWordsShouldBeInvalid() {
+    String notAWord = "This is not a word";
+    String notAWordButSame = notAWord;
+
+    assertEquals(checker.check(notAWord, notAWordButSame), false, "Sentences aren't words.");
+  }
+
+  @Test
+  public void stringsWithSpecialCharactersShouldBeInvalid() {
```
Comment (haixiaosu) [inline elsatellite/assignment-1/src/test/AnagramCheckerTest.java]: Although I think your solution handles this case?

File: elsatellite/assignment-1/src/test/AnagramCheckerTest.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    String notAWord = ">:(";
+    String notAWordButSame = notAWord;
+
+    assertEquals(checker.check(notAWord, notAWordButSame), false, "Special characters aren't words.");
+  }
+
+  @Test
+  public void stringsWithNumberShouldBeInvalid() {
```
Comment (haixiaosu) [inline elsatellite/assignment-1/src/test/AnagramCheckerTest.java]: Ditto here, your solution probably handles this too?

File: elsatellite/assignment-1/src/main/AnagramChecker.java (hunk 3 above)
Comment (elsatellite) [reply elsatellite/assignment-1/src/main/AnagramChecker.java]: The assignment said `Two words are anagrams of each other if they are made of the same letters in a different order` I thought that `letters` was an important word here - maybe it's not?

File: elsatellite/assignment-1/src/main/AnagramChecker.java (hunk 4 above)
Comment (elsatellite) [reply elsatellite/assignment-1/src/main/AnagramChecker.java]: I assumed that the main assignment was about the dictionary meaning of "anagram", in a lexical sense. It says `made of the same letters in a different order`. To me, `a` and `A` are `the same letters`.
But you're right, if we only deal with lowercase, it's not needed at all!

File: elsatellite/assignment-1/src/test/AnagramCheckerTest.java (hunk 5 above)
Comment (elsatellite) [reply elsatellite/assignment-1/src/test/AnagramCheckerTest.java]: The "solution" does, but the checks before make sure that it doesn't return true for strings that aren't words. I tried to stick with my assumptions about the meaning of `letters`.

I don't know if it was the right call :)

**Comment by elsatellite** [comment]
Thanks for the review and your time! 🙌
I might have over-assumed about the directives of the assignment.
And thank you for the advice about the permutations tests, that sounds right, I will do the changes asap!
